crates/services/sync/src/import.rs

```diff
@@ -516,27 +563,47 @@ where
 }
 
 async fn get_transactions<P>(
-    peer_id: PeerId,
     range: Range<u32>,
+    peer_id: Option<PeerId>,
     p2p: &Arc<P>,
-) -> Option<Vec<Transactions>>
+) -> Option<SourcePeer<Vec<Transactions>>>
 where
     P: PeerToPeerPort + Send + Sync + 'static,
 {
-    let range = peer_id.clone().bind(range);
-    let res = p2p
-        .get_transactions(range)
-        .await
-        .trace_err("Failed to get transactions");
-    match res {
-        Ok(Some(transactions)) => Some(transactions),
-        _ => {
-            report_peer(
-                p2p,
-                Some(peer_id.clone()),
-                PeerReportReason::MissingTransactions,
-            );
-            None
+    match peer_id {
+        Some(peer_id) => {
+            let source_peer = peer_id.clone().bind(range.clone());
+            let Ok(Some(txs)) = p2p
+                .get_transactions_from_peer(source_peer)
+                .await
+                .trace_err("Failed to get transactions")
+            else {
+                report_peer(
+                    p2p,
+                    Some(peer_id.clone()),
+                    PeerReportReason::MissingTransactions,
+                );
+                return None;
+            };
+            Some(SourcePeer { peer_id, data: txs })
+        }
```

Comment:
I think so because if we are in the case where we don't use cache and we already fetched the header to a particular peer and we have his peer_id it's more optimize to directly asks him the transactions instead of running computation to find someone with these infos (and probably end-up on him also)